app/models/account.rb

```diff
@@ -11,27 +11,110 @@ class Account < ApplicationRecord
+    accounts_with_period_balances = AccountBalance.with(
+      ranked_balances: ranked_balances_cte
     )
+      .from("ranked_balances AS rb")
+      .joins("JOIN accounts a ON a.id = rb.account_id")
+      .select("
+        a.name,
+        a.accountable_type,
+        a.classification,
+        SUM(CASE WHEN rb.rn_asc = 1 THEN rb.balance ELSE 0 END) AS start_balance,
+        MAX(CASE WHEN rb.rn_asc = 1 THEN rb.date ELSE NULL END) as start_date,
+        SUM(CASE WHEN rb.rn_desc = 1 THEN rb.balance ELSE 0 END) AS end_balance,
+        MAX(CASE WHEN rb.rn_desc = 1 THEN rb.date ELSE NULL END) as end_date
+      ")
+      .where("rb.rn_asc = 1 OR rb.rn_desc = 1")
+      .group("a.id")
+      .order("end_balance")
+      .to_a
+
+    assets = accounts_with_period_balances.select { |row| row.classification == "asset" }
+    liabilities = accounts_with_period_balances.select { |row| row.classification == "liability" }
+
+    total_assets = assets.sum(&:end_balance)
+    total_liabilities = liabilities.sum(&:end_balance)
+
+    {
+      asset: {
+        total: total_assets,
+        groups: assets.group_by(&:accountable_type).transform_values do |rows|
+          end_balance = rows.sum(&:end_balance)
+          start_balance = rows.sum(&:start_balance)
+          {
+            start_balance: start_balance,
+            end_balance: end_balance,
+            allocation: (end_balance / total_assets * 100).round(2),
+            trend: Trend.new(current: end_balance, previous: start_balance, type: "asset"),
+            accounts: rows.map do |account|
+              {
+                name: account.name,
+                start_balance: account.start_balance,
+                end_balance: account.end_balance,
+                allocation: (account.end_balance / total_assets * 100).round(2),
+                trend: Trend.new(current: account.end_balance, previous: account.start_balance, type: "asset")
+              }
+            end
+          }
+        end
+      },
+      liability: {
+        total: total_liabilities,
+        groups: liabilities.group_by(&:accountable_type).transform_values do |rows|
+          end_balance = rows.sum(&:end_balance)
+          start_balance = rows.sum(&:start_balance)
+          {
+            start_balance: start_balance,
+            end_balance: end_balance,
+            allocation: (end_balance / total_liabilities * 100).round(2),
+            trend: Trend.new(current: end_balance, previous: start_balance, type: "liability"),
+            accounts: rows.map do |account|
+              {
+                name: account.name,
+                start_balance: account.start_balance,
+                end_balance: account.end_balance,
+                allocation: (account.end_balance / total_liabilities * 100).round(2),
+                trend: Trend.new(current: account.end_balance, previous: account.start_balance, type: "liability")
+              }
+            end
+          }
+        end
+      }
+    }
   end
```

Comment:
Plenty of refactoring / organizing to do here still—I decided to leave in its current state so that a separate PR can be opened to specifically address a refactor.